util/encoding.go

```diff
@@ -0,0 +1,243 @@
+package util
+
+import (
+	"encoding/base64"
+	"errors"
+	"unsafe"
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = errors.New("the number to be encoded is out of range")
+	fibEncodeInvalidRange     = errors.New("the range is invalid")
+)
+
+func getByteSlice() []byte {
+	// Most string to be encoded are less than 8 bytes.
+	return make([]byte, 0, 8)
+}
+
+func NewBitStreamForWrite() *BitStream {
+	return NewBitStream(getByteSlice())
+}
+
+// enlarge the underlying byte slice.
+// In short, if the final size is less than the expThreshold, double the size of the underlying byte array,
+// otherwise, enlarge the size of the array by a specified step.
+func (bs *BitStream) enlarge(n uint16) {
+	final := int(bs.p+n+7) / 8
+	// Under most circumstances, the value of final will hit this if condition.
+	if final <= cap(bs.b) {
+		bs.b = bs.b[:final]
+		return
+	}
+
+	const expThreshold = 64
+	const step = 16
+	var temp []byte
+
+	if final <= expThreshold {
+		finalF := float32(final - 1)
+		// Find out the leftmost one's position.
+		pos := *(*int)(unsafe.Pointer(&finalF))>>23&255 - 127
```

Comment:
I am not sure what this is calculating or what expThreshold is supposed to be?

Reply:
Would it be better to say `size := cap(bs.b) * 2` rather than using this complex function that is creating an int pointer out of a float pointer? This would seem to depend strongly on just how the CPU architecture represents ints and floats in memory. 

Actually you would need to loop in case you need more than one doubling in size.
```suggestion
		if final <= expThreshold {
		        size := expThreshold
		        for size > final*2 {
		                size = size/2
		        }
		        temp = make([]byte, final, size)
```
This would also fix the bitstream starting at some odd size.